luarules/gadgets/unit_collision_damage_behavior.lua

```diff
@@ -0,0 +1,171 @@
+local spGetUnitHealth = Spring.GetUnitHealth
+local spGetUnitVelocity = Spring.GetUnitVelocity
+local spSetUnitVelocity = Spring.SetUnitVelocity
+local mathMin = math.min
+local mathMax = math.max
+
+local fallDamageMultipliers = {}
+local unitsMaxImpulse = {}
+local weaponDefIDImpulses = {}
+local transportedUnits = {}
+local unitMasses = {}
+local weaponDefIgnored = {}
+local unitInertiaCheckFlags = {}
+local gameFrame = 0
+local velocityWatchDuration = 8
+
+for unitDefID, unitDef in ipairs(UnitDefs) do
+	local fallDamageMultiplier = unitDef.customParams.fall_damage_multiplier or 1.0
+	fallDamageMultipliers[unitDefID] = fallDamageMultiplier * fallDamageMagnificationFactor
+	unitsMaxImpulse[unitDefID] = unitDef.mass * maxImpulseMultiplier
+	unitMasses[unitDefID] = unitDef.mass
+end
+
+for name, weaponDefID in pairs(Game.envDamageTypes or {}) do
+	weaponDefIgnored[weaponDefID] = true
+end
+
+for weaponDefID, wDef in ipairs(WeaponDefs) do
+	if wDef.damages and wDef.damages.impulseBoost and wDef.damages.impulseFactor then
+		weaponDefIDImpulses[weaponDefID] = {impulseBoost = wDef.damages.impulseBoost, impulseFactor = wDef.damages.impulseFactor}
+	end
+
+	
+	local function maxDamage(damages)
+		local damage = damages[0]
+		for i = 1, #damages do
+			damage = mathMax(damages[i], damage)
+		end
+		return damage
+	end
+	
+	--generate list of exempted weapons to improve performance
+	if wDef.damages and wDef.damages.impulseFactor == 0 or
+		(wDef.damages.impulseFactor < minImpulseFactor and wDef.damages.impulseBoost < maxDamage(wDef.damages) * minImpulseToDamageRatio) then
+		weaponDefIgnored[weaponDefID] = true
+	end
+end
+
+local function getImpulseMultiplier(unitDefID, weaponDefID, damage)
+	local impulseBoost = weaponDefIDImpulses[weaponDefID].impulseBoost or 0
+	local impulseFactor = weaponDefIDImpulses[weaponDefID].impulseFactor or 1
+	local impulse = (damage + impulseBoost) * impulseFactor
+	local maxImpulse = unitsMaxImpulse[unitDefID]
+	local impulseMultiplier = mathMin(maxImpulse/impulse, 1)
+	return impulseMultiplier
+end
+
+local function massToHealthRatioMultiplier(unitID, unitDefID)
+	local health, maxHealth = spGetUnitHealth(unitID)
+	if maxHealth then
+		local massToMaxHealthMultiplier = (maxHealth / unitMasses[unitDefID]) * fallDamageMultipliers[unitDefID]
+		return massToMaxHealthMultiplier
+	else
+		return fallDamageMultipliers[unitDefID]
+	end
+end
+
+local function isValidCollisionDirection(unitID)
+	local velX, velY, velZ, velLength = spGetUnitVelocity(unitID)
+		-- y-velocity check prevents mostly horizontal object collisions from taking damage, allows damage if dropped from above
+		return velLength > collisionVelocityThreshold and -velY > (velLength * 0.82)
+end
+
+function gadget:UnitPreDamaged(unitID, unitDefID, unitTeam, damage, paralyzer, weaponDefID, projectileID, attackerID, attackerDefID, attackerTeam)
+	if not weaponDefIgnored[weaponDefID] and weaponDefID > 0 then --this section handles limiting maximum impulse
```

Comment:
`>= 0` shows intent more clearly

Reply:
0 is currently some game-defined weapon and may have impulse (try `Spring.Echo(WeaponDefs[0].name)`). This will change in some future engine version but so far it should use `>= 0`.

Reply:
Done, thanks brother